### PR TITLE
Implement type normalization

### PR DESCRIPTION
**Pull request by alessionossa**
Implement basic type normalization, as discussed in #7.

### CODE REVIEW

**Comment by stackotter** [comment]
Hey Alessio, happy to review this PR but I've just been holding off cause it's still marked as a draft. Just letting you know in case that was an accident 👍

File: Sources/MacroToolkit/Type.swift [hunk 1]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            let normalizedType: NormalizedType = .simple(.init(.init(
+                leadingTrivia: type._baseSyntax.leadingTrivia,
+                name: .identifier("AnyObject"),
+                trailingTrivia: type._baseSyntax.trailingTrivia
+            )))
+            return normalizedType
+            
+        case .composition(let type):
+            // Looks like there can only be simple types in composition, with no generics, and therefore we
+            // don't ned to recursively normalize
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:200]: typo of need

remove the blank line after the comment

File: Sources/MacroToolkit/Type.swift [hunk 2]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            
+            var base = "Array<\(arrayTypeSyntax.element)>"
+            if let attributedTypeSyntax {
+                base = base.addingAttributes(from: attributedTypeSyntax)
+            }
+            return NormalizedType(stringLiteral: base)
+        case .classRestriction(let type):
+            // Not handling `_attributedSyntax` because `classRestriction` cannot have any attribute
+            
+            // let normalizedType: NormalizedType = "AnyObject"
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:190]: Remove this commented out line and then the blank line above it too

File: Sources/MacroToolkit/Type.swift [hunk 3]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            var attributedTypeSyntax: AttributedTypeSyntax? = type._attributedSyntax
+            let normalizedBaseType = Type(type._baseSyntax.baseType).normalized()
+            
+            memberTypeSyntax.genericArgumentClause = memberTypeSyntax.genericArgumentClause?.normalized()
+            
+            memberTypeSyntax.baseType = TypeSyntax(normalizedBaseType._syntax)
+            attributedTypeSyntax?.baseType = TypeSyntax(memberTypeSyntax)
+            
+            return .member(.init(memberTypeSyntax, attributedSyntax: attributedTypeSyntax))
+
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:272]: Remove blank lines after the end of case blocks.

File: Sources/MacroToolkit/Type.swift [hunk 4]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            attributedTypeSyntax?.baseType = TypeSyntax(optionalTypeSyntax)
+            
+//            let identifierSyntax = IdentifierTypeSyntax(
+//                leadingTrivia: optionalTypeSyntax.leadingTrivia,
+//                name: .identifier("Optional"),
+//                genericArgumentClause: .init(
+//                    arguments: .init(arrayLiteral: .init(argument: optionalTypeSyntax.wrappedType))
+//                ),
+//                trailingTrivia: optionalTypeSyntax.leadingTrivia
+//            )
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:303]: Why doesn't this approach work? It'd be preferable but I assume there was something about it that caused you to try the approach below instead.

File: Sources/MacroToolkit/NormalizedType.swift [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                .function(let type as any TypeProtocol),
+                .implicitlyUnwrappedOptional(let type as any TypeProtocol),
+                .member(let type as any TypeProtocol),
+                .missing(let type as any TypeProtocol),
+                .packExpansion(let type as any TypeProtocol),
+                .packReference(let type as any TypeProtocol),
+                .simple(let type as any TypeProtocol),
+                .suppressed(let type as any TypeProtocol),
+                .tuple(let type as any TypeProtocol):
+            type
```
Comment (stackotter) [inline Sources/MacroToolkit/NormalizedType.swift:42]: Indent this case block by one more level (just a stylistic choice I've made for the rest of the repo)

File: Sources/MacroToolkit/NormalizedTypes/NormalizedCompositionType.swift [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import SwiftSyntax
+
+/// Wraps a composition type (e.g. `ProtocolA & ProtocolB`).
+public struct NormalizedCompositionType: TypeProtocol {
```
Comment (stackotter) [inline Sources/MacroToolkit/NormalizedTypes/NormalizedCompositionType.swift:4]: Add a way to access the types that have been composed. Even though they don't need to be recursively normalized, I think it'd still be good to represent them as `NormalizedSimpleType`s so that they can be used as if they've been normalized.

File: Sources/MacroToolkit/NormalizedTypes/NormalizedMemberType.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import SwiftSyntax
+
+/// Wraps a member type (e.g. `Array<Int>.Element`).
+public struct NormalizedMemberType: TypeProtocol {
```
Comment (stackotter) [inline Sources/MacroToolkit/NormalizedTypes/NormalizedMemberType.swift:4]: Add a way to access the root and member types (as normalized types). Same goes for any other normalized types which don't have such properties yet (they should have a simple way to pull out their normalized component parts).

File: Sources/MacroToolkit/Type.swift [hunk 8]
```diff
@@ -135,25 +135,20 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
-            } else if tupleSyntax.elements.count == 1, let element = tupleSyntax.elements.first {
-                // TODO: Can we assume that we won't get a single-element tuple with a label (which would be invalid anyway)?
-                return element.type.withoutTrivia().description
-            } else {
-                return _syntax.withoutTrivia().description
-            }
-        } else {
-            return _syntax.withoutTrivia().description
-        }
+        self.normalized()._syntax.withoutTrivia().description
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:138]: Remove unnecessary `self`

File: Sources/MacroToolkit/Type.swift [hunk 9]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            // let normalizedType: NormalizedType = "AnyObject"
+            let normalizedType: NormalizedType = .simple(.init(.init(
+                leadingTrivia: type._baseSyntax.leadingTrivia,
+                name: .identifier("AnyObject"),
+                trailingTrivia: type._baseSyntax.trailingTrivia
+            )))
+            return normalizedType
+            
+        case .composition(let type):
+            // Looks like there can only be simple types in composition, with no generics, and therefore we
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:199]: How is syntax such as `(Encodable & Sequence<Int?>) & Equatable` parsed? I assume it gets interpreted as something like `(Encodable & Sequence<Int?>) & Equatable` in which case you still need to continue recursively normalizing to reach `Int?`.

File: Sources/MacroToolkit/Type.swift [hunk 10]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            
+            return .packExpansion(.init(type._baseSyntax, attributedSyntax: type._attributedSyntax))
+        case .packReference(let type):
+            // Looks like there can only be simple identifiers in pack references, with no generics, and therefore we
+            // don't ned to recursively normalize
+            
+            return .packReference(.init(type._baseSyntax, attributedSyntax: type._attributedSyntax))
+        case .simple(let type):
+            if type.name == "Void" {
+                return .tuple(.init(.init(elements: [])))
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:322]: I think `Void` is better to have as a simple type as that's the name most people know it by (and using `()` is considered less swift-y by a few core Swift contributors, even though `Void` is actually an alias to `()` under-the-hood).

File: Sources/MacroToolkit/Type.swift [hunk 11]
```diff
@@ -174,6 +169,199 @@ public enum `Type`: TypeProtocol, SyntaxExpressibleByStringInterpolation {
...
+            var suppressedTypeSyntax: SuppressedTypeSyntax = type._baseSyntax
+            
+            let normalizedConstraint = Type(suppressedTypeSyntax.type).normalized()
+            suppressedTypeSyntax.type = TypeSyntax(normalizedConstraint._syntax)
+                        
+            return .suppressed(.init(suppressedTypeSyntax))
+        case .tuple(let type):
+            if type.elements.count == 1 {
+                let child = type.elements[0]
+                if case .tuple(_) = child {
```
Comment (stackotter) [inline Sources/MacroToolkit/Type.swift:344]: I believe you can do `if case .tuple = child` which is a little bit cleaner, but it doesn't really matter either way.

File: Tests/MacroToolkitTests/MacroToolkitTests.swift [hunk 12]
```diff
@@ -678,4 +678,221 @@ final class MacroToolkitTests: XCTestCase {
...
+
+        XCTAssertEqual(normalizedType.description, "()")
+    }
+    
+    func testNormalizationMultipleVoidWithOptional() {
+        let type: `Type` = "(((((()))?)))"
+
+        let normalizedType = type.normalized()
+
+        XCTAssertEqual(normalizedType.description, "(Optional<()>)")
```
Comment (stackotter) [inline Tests/MacroToolkitTests/MacroToolkitTests.swift:717]: The outer layer of brackets should get removed by normalization since `(Optional<()>)` is equivalent to `Optional<()>`.

Comment (stackotter) [reply Tests/MacroToolkitTests/MacroToolkitTests.swift:717]: But to complicate things, the brackets in `((Int) -> (Int)) -> String` can't be removed (a bit of an annoying edge case to deal with). I guess the rule might be that if the brackets are around a function type which itself is the input type of another function type then they can't be omitted?

Comment (stackotter) [reply Tests/MacroToolkitTests/MacroToolkitTests.swift:717]: Oh, and if the brackets are around an implicitly unwrapped optional function type they can't be removed either.

File: Tests/MacroToolkitTests/MacroToolkitTests.swift [hunk 13]
```diff
@@ -678,4 +678,221 @@ final class MacroToolkitTests: XCTestCase {
...
+        let type: `Type` = "~Copyable"
+                
+        let normalizedType = type.normalized()
+
+        XCTAssertEqual(normalizedType.description, "~Copyable")
+    }
+    
+    func testNormalizationAttributedTuple() {
+        let declSyntax: DeclSyntax = """
+        let interestingType: (inout [Int], String) -> Float = { _,_ in
```
Comment (stackotter) [inline Tests/MacroToolkitTests/MacroToolkitTests.swift:879]: Add a space after the comma in `_, _` to match the repo's code style (even thought it's not actually code).

**Review by stackotter — CHANGES_REQUESTED** [review]
Looking awesome so far! 🚀 It seems like you've caught most of the edge cases 👌 I've left some comments with a few more edge cases to test out and some aspects of normalization that I think should be changed to make things a bit easier to use or more intuitive. If there's anything that doesn't make sense or you don't agree with, don't hesitate to let me know! Always happy to have a more in-depth discussion.

**Comment by stackotter** [comment]
> As a source of truth for what the normalized version should be, I used, where possible, the result obtained by printing the type to the console, for example:
> 
> ...
> 
> ```swift
> print("Void: \(Void.self)")    // Void: ()
> ```

Hmm that does complicate my suggestion for `Void` to be normalized to a simple type a bit. I still reckon in this case we should go with `Void` as the normalized representation since that's what people know it by and what most people prefer in their own code, but being able to use the string representations of types as the source of truth for normalization is a very compelling argument against that idea. I'd love to hear your thoughts.

File: Sources/MacroToolkit/NormalizedTypes/NormalizedCompositionType.swift (hunk 6 above)
Comment (alessionossa) [reply Sources/MacroToolkit/NormalizedTypes/NormalizedCompositionType.swift:4]: I based the implementation of `NormalizedCompositionType` on `CompositionType`, that has no direct way to access the sub-types. Should I still implement this property on `NormalizedCompositionType`?

File: Sources/MacroToolkit/Type.swift (hunk 1 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: Fixed in 71aff30.

File: Sources/MacroToolkit/Type.swift (hunk 2 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: Fixed in 71aff30.

File: Sources/MacroToolkit/Type.swift (hunk 3 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: Fixed in 71aff30.

File: Sources/MacroToolkit/NormalizedType.swift (hunk 5 above)
Comment (alessionossa) [reply Sources/MacroToolkit/NormalizedType.swift]: Not the standard indentation, but fixed in 71aff30.

File: Sources/MacroToolkit/Type.swift (hunk 8 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: Fixed in 71aff30.

File: Tests/MacroToolkitTests/MacroToolkitTests.swift (hunk 13 above)
Comment (alessionossa) [reply Tests/MacroToolkitTests/MacroToolkitTests.swift]: Fixed in 84c2bd6.

File: Sources/MacroToolkit/Type.swift (hunk 11 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: Fixed with the updated implementation in b4a470c.

File: Sources/MacroToolkit/Type.swift (hunk 4 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift]: This approach does actually work, the other one was just less code and most likely was as effective 99% of the time. I now switched to this more verbose approach that maintains trivia information with 206fe0a.

File: Sources/MacroToolkit/NormalizedTypes/NormalizedMemberType.swift (hunk 7 above)
Comment (alessionossa) [reply Sources/MacroToolkit/NormalizedTypes/NormalizedMemberType.swift:4]: Same as https://github.com/stackotter/swift-macro-toolkit/pull/15#discussion_r1684684292 above, `MemberType` does not have any direct way to access such properties.

File: Tests/MacroToolkitTests/MacroToolkitTests.swift (hunk 12 above)
Comment (alessionossa) [reply Tests/MacroToolkitTests/MacroToolkitTests.swift:717]: In b4a470c I updated the implementation of tuple normalization to remove brackets around simple types too. This allows for both of `(Optional<()>)` and `((Int) -> (Int)) -> String` to be normalized correctly. 
In the tests, I added a check against the string representations of types, on top of the hardcoded expected nromalization, to be sure we are aligned with how Swift compiler normalize those types.

**Comment by alessionossa** [comment]
> Hmm that does complicate my suggestion for `Void` to be normalized to a simple type a bit. I still reckon in this case we should go with `Void` as the normalized representation since that's what people know it by and what most people prefer in their own code, but being able to use the string representations of types as the source of truth for normalization is a very compelling argument against that idea. I'd love to hear your thoughts.

I found useful to use string representations of types in my code to avoid misspelling types and keeps cases as type-safe as possible.

In my code using the package I have, for example:
```swift
switch (typeName) {
case ("\(Int.self)"):
    getterBody = "integer(forKey: Self.\(identifierTokenSyntax)Key)"
case ("\(Bool.self)"):
    getterBody = "bool(forKey: Self.\(identifierTokenSyntax)Key)"
case ("\([String: Any].self)", true):
    getterBody = "dictionary(forKey: Self.\(identifierTokenSyntax)Key)"
```

Even considering that `Void` is generally preferable over `()` in Swift, we are here trying to normalize types to their most basic form, not trying to make them look nice and swift-y. Also `Int?` is preferable over `Optional<Int>`, but `Int?` it’s not the normalized representation of it.

File: Sources/MacroToolkit/Type.swift (hunk 10 above)
Comment (alessionossa) [reply Sources/MacroToolkit/Type.swift:326]: See https://github.com/stackotter/swift-macro-toolkit/pull/15#issuecomment-2241100094.

**Comment by stackotter** [comment]
> Even considering that `Void` is generally preferable over `()` in Swift, we are here trying to normalize types to their most basic form, not trying to make them look nice and swift-y. Also `Int?` is preferable over `Optional<Int>`, but `Int?` it’s not the normalized representation of it.

Yeah that is a tricky distinction to make. The way that I'm looking at it is that the two concurrent goals for the normalisation are to minimise the number of special cases (i.e. removing specialised `Int?` syntax etc) and to make the normalised form as predictable as possible (so that developers don't incorrectly assume how something will get normalised).

As I see it, normalisation now has only a single way of representing every type (ignoring the possibility for unnecessary single-element tuple syntax), except for void; macro toolkit has to make a choice between either `.simple(.init("Void"))` or `.tuple(.init(.init(elements: [])))`.

To me, comparing a type against `.simple(.init("Void"))` makes it immediately obvious what you're checking for. Whereas it'd be harder to skim-read `.tuple(.init(.init(elements: [])))` since you'd usually initially assume that it's an actual tuple not `()` which many people would probably pronounce as 'void' when reading code such as `(Int) -> ()`.

If choosing one option or the other eliminated an enum case from `NormalizedType` then it'd be a pretty easy choice but unfortunately neither actually simplifies the model so it's more just a question of predictability. I don't think either option's perfect, but I'm still leaning heavily towards `Void` in terms of how much more readable it is in user code. The only thing that's making me consider `()` at all is the fact that Swift prints void as `()`.

I guess if you look at it from the 'reducing the number of special cases' point of view then representing void as `"Void"` does help reduce the number of special cases of `.tuple`, because at the moment it can have any number of elements greater than 1, or it can have zero elements. Whereas if void is `"Void"` then `.tuple` quite simply just requires more than 1 element.

I'll attempt to do a bit of polling among Swifties that I know and see what people reckon.

**Comment by alessionossa** [comment]
@stackotter Is there any update on this?
